File: app/src/main/java/com/esbati/keivan/persiancalendar/components/views/TextCheckCell.java
/*
 * This is the source code of Telegram for Android v. 3.x.x.
 * It is licensed under GNU GPL v. 2 or later.
 * You should have received a copy of the license in this archive (see LICENSE).
 *
 * Copyright Nikolai Kudashov, 2013-2016.
 */

package com.esbati.keivan.persiancalendar.components.views;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.os.Build;
import android.support.v7.widget.SwitchCompat;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.TextView;

import com.esbati.keivan.persiancalendar.R;
import com.esbati.keivan.persiancalendar.utils.LayoutHelper;

import static com.esbati.keivan.persiancalendar.utils.AndroidUtilitiesKt.toDp;

public class TextCheckCell extends FrameLayout {

    private TextView textView;
    private SwitchCompat checkBox;
    private static Paint paint;
    private boolean needDivider;

    public TextCheckCell(Context context) {
        super(context);

        if (paint == null) {
            paint = new Paint();
            paint.setColor(0xffd9d9d9);
            paint.setStrokeWidth(1);
        }

        textView = new TextView(context);
        textView.setTextColor(getResources().getColorStateList(R.color.text_setting_key));
        textView.setTextSize(TypedValue.COMPLEX_UNIT_DIP, 16);
        textView.setLines(1);
        textView.setMaxLines(1);
        textView.setSingleLine(true);
        textView.setGravity(Gravity.RIGHT | Gravity.CENTER_VERTICAL);
        addView(textView, LayoutHelper.createFrame(LayoutHelper.MATCH_PARENT, LayoutHelper.MATCH_PARENT, Gravity.RIGHT | Gravity.TOP, 17, 0, 17, 0));

        checkBox = new SwitchCompat(context);
        checkBox.setDuplicateParentStateEnabled(false);
        checkBox.setFocusable(false);
        checkBox.setFocusableInTouchMode(false);
        checkBox.setClickable(false);
        addView(checkBox, LayoutHelper.createFrame(LayoutHelper.WRAP_CONTENT, LayoutHelper.WRAP_CONTENT, Gravity.LEFT | Gravity.CENTER_VERTICAL, 14, 0, 14, 0));
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, View.MeasureSpec.makeMeasureSpec(toDp(48) + (needDivider ? 1 : 0), View.MeasureSpec.EXACTLY));
    }

    public void setTextAndCheck(String text, boolean checked, boolean divider) {
        textView.setText(text);
        if (Build.VERSION.SDK_INT < 11) {
            checkBox.requestLayout();
        }
        checkBox.setChecked(checked);
        needDivider = divider;
        setWillNotDraw(!divider);
    }

    public void setChecked(boolean checked) {
        checkBox.setChecked(checked);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        if (needDivider) {
            canvas.drawLine(getPaddingLeft(), getHeight() - 1, getWidth() - getPaddingRight(), getHeight() - 1, paint);
        }
    }

    @Override
    public void setEnabled(boolean isEnabled){
        super.setEnabled(isEnabled);
        for(int i = 0 ; i < getChildCount() ; i++)
            getChildAt(i).setEnabled(isEnabled);
    }
}

File: app/src/main/java/com/esbati/keivan/persiancalendar/features/settings/cells/TextCheckCell.kt
package com.esbati.keivan.persiancalendar.features.settings.cells

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import androidx.appcompat.widget.SwitchCompat
import android.util.TypedValue
import android.view.Gravity
import android.view.View
import android.widget.FrameLayout
import android.widget.TextView
import com.esbati.keivan.persiancalendar.R
import com.esbati.keivan.persiancalendar.features.settings.ViewMarker
import com.esbati.keivan.persiancalendar.utils.LayoutHelper
import com.esbati.keivan.persiancalendar.utils.toDp

@ViewMarker
class TextCheckCell(context: Context) : FrameLayout(context) {

    private val textView: TextView
    private val checkBox: SwitchCompat
    var title: String
        get() = textView.text.toString()
        set(value) {
            textView.text = value
        }
    var isChecked: Boolean
        get() = checkBox.isChecked
        set(value) {
            checkBox.isChecked = value
        }
    var needDivider: Boolean = false
        set(value) {
            field = value
            setWillNotDraw(!value)
        }


    init {
        textView = TextView(context).apply {
            setTextColor(resources.getColorStateList(R.color.text_setting_key))
            setTextSize(TypedValue.COMPLEX_UNIT_DIP, 16f)
            setLines(1)
            maxLines = 1
            setSingleLine(true)
            gravity = Gravity.RIGHT or Gravity.CENTER_VERTICAL
        }
        addView(textView, LayoutHelper.createFrame(
                LayoutHelper.MATCH_PARENT, LayoutHelper.MATCH_PARENT.toFloat()
                , Gravity.RIGHT or Gravity.TOP
                , 17f, 0f, 17f, 0f)
        )

        checkBox = SwitchCompat(context).apply {
            isDuplicateParentStateEnabled = false
            isFocusable = false
            isFocusableInTouchMode = false
            isClickable = false
        }
        addView(checkBox, LayoutHelper.createFrame(
                LayoutHelper.WRAP_CONTENT, LayoutHelper.WRAP_CONTENT.toFloat()
                , Gravity.LEFT or Gravity.CENTER_VERTICAL
                , 14f, 0f, 14f, 0f)
        )
    }

    fun setTextAndCheck(text: String, checked: Boolean, divider: Boolean) {
        textView.text = text
        checkBox.isChecked = checked
        needDivider = divider
        setWillNotDraw(!divider)
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, View.MeasureSpec.makeMeasureSpec(
                48.toDp() + if (needDivider) 1 else 0, View.MeasureSpec.EXACTLY
        ))
    }

    override fun onDraw(canvas: Canvas) {
        if (needDivider)
            canvas.drawLine(
                    paddingLeft.toFloat()
                    , (height - 1).toFloat()
                    , (width - paddingRight).toFloat()
                    , (height - 1).toFloat()
                    , paint
            )
    }

    override fun setEnabled(isEnabled: Boolean) {
        super.setEnabled(isEnabled)
        for (i in 0 until childCount)
            getChildAt(i).isEnabled = isEnabled
    }

    companion object {
        private val paint: Paint by lazy {
            Paint().apply {
                color = -0x262627
                strokeWidth = 1f
            }
        }
    }
}
